File: scripts/test-streams.py
#!/usr/bin/env python3

import urllib.request
import urllib.error
import socket
import sys
from pathlib import Path
import json
from concurrent.futures import ThreadPoolExecutor, as_completed
import time

M3U_FILE = Path(__file__).parent.parent / "Fifa world cup.m3u"

def parse_m3u(filepath):
    """Parse M3U file and return list of channels"""
    channels = []
    with open(filepath, 'r', encoding='utf-8') as f:
        lines = f.readlines()
    
    current_name = ""
    for line in lines:
        line = line.strip()
        if line.startswith("#EXTINF:-1"):
            # Extract channel name
            if "," in line:
                current_name = line.split(",", 1)[1]
        elif line and not line.startswith("#") and current_name:
            channels.append({
                "name": current_name,
                "url": line
            })
            current_name = ""
    
    return channels

def test_stream(channel, timeout=2):
    """Test if a stream URL is accessible"""
    try:
        req = urllib.request.Request(
            channel['url'],
            headers={'User-Agent': 'Mozilla/5.0'}
        )
        socket.setdefaulttimeout(timeout)
        response = urllib.request.urlopen(req, timeout=timeout)
        return channel, True
    except:
        return channel, False

def test_all_streams():
    """Test all streams in parallel"""
    channels = parse_m3u(M3U_FILE)
    working = []
    failed = []
    
    print(f"\n🔍 Testing {len(channels)} channels...\n")
    
    with ThreadPoolExecutor(max_workers=10) as executor:
        futures = {executor.submit(test_stream, ch): idx for idx, ch in enumerate(channels)}
        
        for idx, future in enumerate(as_completed(futures), 1):
            channel, is_working = future.result()
            name = channel['name'][:50].ljust(50)
            status = "✅" if is_working else "❌"
            print(f"[{idx}/{len(channels)}] {name} {status}")
            
            if is_working:
                working.append(channel)
            else:
                failed.append(channel)
    
    print("\n" + "="*70)
    print(f"\n✅ Working: {len(working)} | ❌ Failed: {len(failed)}\n")
    
    # Save results
    results = {
        "timestamp": time.strftime("%Y-%m-%d %H:%M:%S"),
        "total": len(channels),
        "working": len(working),
        "failed": len(failed),
        "workingChannels": working,
        "failedChannels": failed
    }
    
    output_file = Path(__file__).parent / "stream-test-results.json"
    with open(output_file, 'w') as f:
        json.dump(results, f, indent=2)
    
    print(f"📊 Results saved to: scripts/stream-test-results.json")
    print(f"\n📋 Working Channels ({len(working)}):")
    for idx, ch in enumerate(working, 1):
        print(f"   {idx}. {ch['name']}")
    
    return results

if __name__ == "__main__":
    test_all_streams()

File: scripts/test-streams.js
#!/usr/bin/env node

const fs = require("fs");
const path = require("path");
const http = require("http");
const https = require("https");

const M3U_FILE = path.join(__dirname, "../Fifa world cup.m3u");

function parseM3U(filePath) {
  const content = fs.readFileSync(filePath, "utf-8");
  const lines = content.split("\n");
  const channels = [];
  let currentName = "";

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i].trim();

    if (line.startsWith("#EXTINF:-1")) {
      const match = line.match(/,(.+)$/);
      currentName = match ? match[1].trim() : `Channel ${i}`;
    } else if (line && !line.startsWith("#") && currentName) {
      channels.push({
        name: currentName,
        url: line,
      });
      currentName = "";
    }
  }

  return channels;
}

function testStream(url, timeout = 3000) {
  return new Promise((resolve) => {
    const protocol = url.startsWith("https") ? https : http;
    let isResolved = false;

    const timeoutHandle = setTimeout(() => {
      if (!isResolved) {
        isResolved = true;
        resolve(false);
      }
    }, timeout);

    try {
      const request = protocol.get(
        url,
        { timeout, headers: { "User-Agent": "Mozilla/5.0" } },
        (response) => {
          if (!isResolved) {
            isResolved = true;
            clearTimeout(timeoutHandle);
            resolve(response.statusCode >= 200 && response.statusCode < 400);
            response.destroy();
          }
        }
      );

      request.on("error", () => {
        if (!isResolved) {
          isResolved = true;
          clearTimeout(timeoutHandle);
          resolve(false);
        }
      });

      request.on("timeout", () => {
        if (!isResolved) {
          isResolved = true;
          clearTimeout(timeoutHandle);
          request.destroy();
          resolve(false);
        }
      });
    } catch (e) {
      if (!isResolved) {
        isResolved = true;
        clearTimeout(timeoutHandle);
        resolve(false);
      }
    }
  });
}

async function testAllStreams() {
  const channels = parseM3U(M3U_FILE);
  const workingChannels = [];
  const failedChannels = [];

  console.log(`\n🔍 Testing ${channels.length} channels (timeout: 3 seconds per channel)...\n`);

  // Test streams with concurrency limit
  const concurrency = 5;
  for (let i = 0; i < channels.length; i += concurrency) {
    const batch = channels.slice(i, i + concurrency);
    const promises = batch.map(async (channel, idx) => {
      const currentIndex = i + idx + 1;
      process.stdout.write(
        `[${currentIndex}/${channels.length}] ${channel.name.substring(0, 40).padEnd(40)}... `
      );

      const isWorking = await testStream(channel.url);
      console.log(isWorking ? "✅" : "❌");

      return { channel, isWorking };
    });

    const results = await Promise.all(promises);
    results.forEach(({ channel, isWorking }) => {
      if (isWorking) {
        workingChannels.push(channel);
      } else {
        failedChannels.push(channel);
      }
    });
  }

  console.log("\n" + "=".repeat(70));
  console.log(`\n✅ Working: ${workingChannels.length} | ❌ Failed: ${failedChannels.length}\n`);

  // Save results
  const results = {
    timestamp: new Date().toISOString(),
    total: channels.length,
    working: workingChannels.length,
    failed: failedChannels.length,
    workingChannels,
    failedChannels,
  };

  fs.writeFileSync(
    path.join(__dirname, "stream-test-results.json"),
    JSON.stringify(results, null, 2)
  );

  console.log("📊 Results saved to: scripts/stream-test-results.json");
  console.log("\n📋 Working Channels List:");
  workingChannels.forEach((ch, idx) => {
    console.log(`   ${idx + 1}. ${ch.name}`);
  });

  return results;
}

testAllStreams().catch(console.error);
